components/script/dom/cssmediarule.rs

```diff
@@ -14,58 +14,71 @@ use dom::cssstylesheet::CSSStyleSheet;
 }
 
 impl CSSMediaRule {
-    fn new_inherited(parent_stylesheet: &CSSStyleSheet, mediarule: Arc<RwLock<MediaRule>>)
+    fn new_inherited(parent_stylesheet: &CSSStyleSheet, mediarule: Arc<Locked<MediaRule>>)
                      -> CSSMediaRule {
-        let list = mediarule.read().rules.clone();
+        let guard = parent_stylesheet.shared_lock().read();
+        let list = mediarule.read_with(&guard).rules.clone();
         CSSMediaRule {
-            cssrule: CSSConditionRule::new_inherited(parent_stylesheet, list),
+            cssconditionrule: CSSConditionRule::new_inherited(parent_stylesheet, list),
             mediarule: mediarule,
             medialist: MutNullableJS::new(None),
         }
     }
 
     #[allow(unrooted_must_root)]
     pub fn new(window: &Window, parent_stylesheet: &CSSStyleSheet,
-               mediarule: Arc<RwLock<MediaRule>>) -> Root<CSSMediaRule> {
+               mediarule: Arc<Locked<MediaRule>>) -> Root<CSSMediaRule> {
         reflect_dom_object(box CSSMediaRule::new_inherited(parent_stylesheet, mediarule),
                            window,
                            CSSMediaRuleBinding::Wrap)
     }
 
     fn medialist(&self) -> Root<MediaList> {
-        self.medialist.or_init(|| MediaList::new(self.global().as_window(),
-                                                 self.mediarule.read().media_queries.clone()))
+        self.medialist.or_init(|| {
+            let guard = self.cssconditionrule.shared_lock().read();
+            MediaList::new(self.global().as_window(),
+                           self.cssconditionrule.parent_stylesheet(),
+                           self.mediarule.read_with(&guard).media_queries.clone())
+        })
     }
 
     /// https://drafts.csswg.org/css-conditional-3/#the-cssmediarule-interface
     pub fn get_condition_text(&self) -> DOMString {
-        let rule = self.mediarule.read();
-        let list = rule.media_queries.read();
+        let guard = self.cssconditionrule.shared_lock().read();
+        let rule = self.mediarule.read_with(&guard);
+        let list = rule.media_queries.read_with(&guard);
         list.to_css_string().into()
     }
 
     /// https://drafts.csswg.org/css-conditional-3/#the-cssmediarule-interface
     pub fn set_condition_text(&self, text: DOMString) {
         let mut input = Parser::new(&text);
         let new_medialist = parse_media_query_list(&mut input);
-        let rule = self.mediarule.read();
-        let mut list = rule.media_queries.write();
-        *list = new_medialist;
+        let mut guard = self.cssconditionrule.shared_lock().write();
+
+        // Clone an Arc because we can’t borrow `guard` twice at the same time.
+
+        // FIXME(SimonSapin): allow access to multiple objects with one write guard?
+        // Would need a set of usize pointer addresses or something,
+        // the same object is not accessed more than once.
+        let mqs = Arc::clone(&self.mediarule.write_with(&mut guard).media_queries);
```

Comment:
We could avoid this by replacing `Arc<Locked<MediaRule>>` with `Arc<MediaRule>`. This particular lock is not useful since each of `MediaRule`’s two fields is itself `Arc<Locked<_>>`. However doing so would break some macros in `style/stylesheets.rs` and `ports/gecko/glue.rs` that generate similar code (including calls to `Locked::read_with`) for all types of rules. Maybe we should do it anyway.